src/wallet/wallet_tx_builder.cpp

```diff
@@ -0,0 +1,683 @@
+                orchardOutputs,
+                this->maxOrchardActions);
+    }
+
+    // Set the dust threshold so that we can select enough inputs to avoid
+    // creating dust change amounts.
+    CAmount dustThreshold{this->DefaultDustThreshold()};
+
+    // TODO: the set of recipient pools is not quite sufficient information here; we should
+    // probably perform note selection at the same time as we're performing resolved payment
+    // construction above.
+    if (!spendableMut.LimitToAmount(targetAmount, dustThreshold, resolved.GetRecipientPools())) {
+        CAmount changeAmount{spendableMut.Total() - targetAmount};
+        return InvalidFundsError(
+                spendableMut.Total(),
+                changeAmount > 0 && changeAmount < dustThreshold
+                // TODO: we should provide the option for the caller to explicitly
+                // forego change (definitionally an amount below the dust amount)
+                // and send the extra to the recipient or the miner fee to avoid
+                // creating dust change, rather than prohibit them from sending
+                // entirely in this circumstance.
+                // (Daira disagrees, as this could leak information to the recipient)
+                ? InvalidFundsReason(DustThresholdError(dustThreshold, changeAmount))
+                : InvalidFundsReason(InsufficientFundsError(targetAmount)));
+    }
+
+    // When spending transparent coinbase outputs, all inputs must be fully
+    // consumed, and they may only be sent to shielded recipients.
+    if (spendableMut.HasTransparentCoinbase()) {
+        if (spendableMut.Total() != targetAmount) {
+            return ChangeNotAllowedError(spendableMut.Total(), targetAmount);
+        } else if (resolved.HasTransparentRecipient()) {
+            return AddressResolutionError::TransparentRecipientNotAllowed;
+        }
+    }
+
+    if (spendableMut.orchardNoteMetadata.size() > this->maxOrchardActions) {
+        return ExcessOrchardActionsError(
+                ActionSide::Input,
+                spendableMut.orchardNoteMetadata.size(),
+                this->maxOrchardActions);
+    }
+
+    return InputSelection(resolved, anchorHeight);
+}
+
+std::pair<uint256, uint256> WalletTxBuilder::SelectOVKs(
+        const ZTXOSelector& selector,
+        const SpendableInputs& spendable) const
+{
+    uint256 internalOVK;
+    uint256 externalOVK;
+    if (!spendable.orchardNoteMetadata.empty()) {
+        std::optional<OrchardFullViewingKey> fvk;
+        std::visit(match {
+            [&](const UnifiedAddress& ua) {
+                auto ufvk = wallet.GetUFVKForAddress(ua);
+                // This is safe because spending key checks will have ensured that we
+                // have a UFVK corresponding to this address, and Orchard notes will
+                // not have been selected if the UFVK does not contain an Orchard key.
+                fvk = ufvk.value().GetOrchardKey().value();
```

Comment:
```suggestion
                assert(ufvk.has_value());
                auto optFvk = ufvk.value().GetOrchardKey();
                assert(optFvk.has_value());
                fvk = optFvk.value();
```

Reply:
Looking at this, I want to change it a bit more significantly, but wondering if you have arguments for/against these options:
1. fewer asserts (like your suggestions)
```c++
        std::optional<OrchardFullViewingKey> fvk = std::visit(match {
            [&](const UnifiedAddress& ua) {
                auto ufvk = wallet.GetUFVKForAddress(ua);
                assert(ufvk.has_value());
                return ufvk.value().GetOrchardKey();
            },
            ...
            [](const auto&) {
                throw std::runtime_error("SelectOVKs: Selector cannot select Orchard notes.");
            }
        }, selector.GetPattern());
        assert(fvk.has_value());

        internalOVK = fvk.value().ToInternalOutgoingViewingKey();
        externalOVK = fvk.value().ToExternalOutgoingViewingKey();
```
or
2. more precise result type
```c++
        OrchardFullViewingKey fvk = std::visit(match {
            [&](const UnifiedAddress& ua) {
                auto ufvk = wallet.GetUFVKForAddress(ua);
                assert(ufvk.has_value());
                auto fvk = ufvk.value().GetOrchardKey();
                assert(fvk.has_value());
                return fvk.value();
            },
            ...
            [](const auto&) {
                throw std::runtime_error("SelectOVKs: Selector cannot select Orchard notes.");
            }
        }, selector.GetPattern());

        internalOVK = fvk.ToInternalOutgoingViewingKey();
        externalOVK = fvk.ToExternalOutgoingViewingKey();
```
The latter also has the benefit of telling us _which_ pattern type was missing an OrchardKey if we hit the assertion, but it’s more verbose.

Reply:
This question is moot – I have a simpler/safer approach.

Reply:
Ok, 8ebb22c67fcb982853ed8b9b22b405dab15a8011 contains what I think is a better approach to `SelectOVKs` – side-by-side is _maybe_ a better way to view the diff.

/cc @nuttycom